src/chttpd_db.erl

```diff
@@ -1628,6 +1633,44 @@ bulk_get_open_doc_revs1(Db, Props, _, {DocId, Revs, Options}) ->
             {DocId, Results, Options}
     end.
 
+maybe_remove_exceed_docs(DocArray0) ->
+    case config:get_boolean("couchdb", "use_max_document_size", false) of
+        true ->
+            Max = config:get_integer("couchdb", "max_document_size",
+                16777216),
+            {ExceedDocs, DocsArray} = lists:splitwith(fun (Doc) ->
+                exceed_doc_size(Doc, Max)
+            end, DocArray0),
+            ExceedErrs = lists:map (fun ({Doc}) ->
+            DocId = case couch_util:get_value(<<"_id">>, Doc) of
+                undefined -> <<"no id generated since update failed">>;
+                Id0 -> Id0
+            end,
+            Reason = lists:concat(["Document exceeded max_document_size:",
+                " of ", Max, " bytes"]),
+            {[{id, DocId}, {error, <<"request_entity_too_large">>},
+                {reason, ?l2b(Reason)}]}
+            end, ExceedDocs),
+            {ExceedErrs, DocsArray};
+        false ->
+            {[], DocArray0}
+    end.
+
+exceed_doc_size(JsonBody, MaxSize) ->
+    size(term_to_binary(JsonBody)) > MaxSize.
```

Comment:
And we should be using erlang:external_size(Term) which doesn't waste resources trying to create the binary just to get a length.